### DISCUSSION TITLE
Bulk submitting Tax Rules via API

### DISCUSSION THREAD

**Discussion by turismoracing** [swellstores/community#552] [Help] [answered]
I would like to know if there is a way to submit bulk tax rules via API.

Ideally I'd like to submit tax rules for each zip code via API.  But if that's not possible, I'd like to do it for all 50 states.

Is it possible via API?

**Comment by sags95** [comment, marked as answer]
Yes it is possible, all of the tax settings can be modified through the `/settings/taxes` endpoint. You can create a new tax rule with a `POST` request to `/settings/taxes/rules`

```
POST '/settings/taxes/rules',
{
  "id": "bc",
  "name": "BC",
  "priority": null,
  "origin": null,
  "categories": [],
  "shipping": false,
  "product_classes": [
    "standard"
  ],
  "account_classes": [],
  "rates": [
    {
      "name": "GST & PST",
      "country": "CA",
      "rate": 13,
      "state": "BC",
      "zip": null,
      "$locale": {
        "en-US": {
          "name": "GST & PST"
        },
        "fr": {
          "name": ""
        }
      }
    }
  ]
}
```

**Comment by turismoracing** [comment]
What is the full URL that I send the POST request to?

**Reply by sags95** [reply]
With `swell-node` it would be `swell.post('/settings/taxes/rules', {...})`

**Reply by mcconnon12** [reply]
Can you do batch post inside swell at /settings/taxes/rules?

**Reply by ericingram** [reply]
You can set/update all rules in a single request if that's what you mean

**Reply by mcconnon12** [reply]
Hey Eric, when I do this I get error messages. But when I did just one rule it worked. 

Here is the error message I get:

{
  "errors": {
    "rules.id": {
      "message": "Required",
      "code": "REQUIRED"
    },
    "rules.name": {
      "message": "Required",
      "code": "REQUIRED"
    }
  }
}

**Reply by ericingram** [reply]
Can you show the full request?

**Comment by mcconnon12** [comment]
Sure... here's an example. Thanks in advance for your help. 

[
    {
        "id": "34487",
        "name": "34487",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34487",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34487"
            }
        ]
    },
    {
        "id": "34465",
        "name": "34465",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34465",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34465"
            }
        ]
    },
    {
        "id": "34464",
        "name": "34464",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34464",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34464"
            }
        ]
    },
    {
        "id": "34460",
        "name": "34460",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34460",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34460"
            }
        ]
    },
    {
        "id": "34453",
        "name": "34453",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34453",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34453"
            }
        ]
    },
    {
        "id": "34452",
        "name": "34452",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34452",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34452"
            }
        ]
    },
    {
        "id": "34451",
        "name": "34451",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34451",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34451"
            }
        ]
    },
    {
        "id": "34450",
        "name": "34450",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34450",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34450"
            }
        ]
    },
    {
        "id": "34448",
        "name": "34448",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34448",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34448"
            }
        ]
    },
    {
        "id": "34447",
        "name": "34447",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34447",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34447"
            }
        ]
    },
    {
        "id": "34445",
        "name": "34445",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34445",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34445"
            }
        ]
    },
    {
        "id": "34442",
        "name": "34442",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34442",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34442"
            }
        ]
    },
    {
        "id": "34441",
        "name": "34441",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34441",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34441"
            }
        ]
    },
    {
        "id": "34434",
        "name": "34434",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34434",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34434"
            }
        ]
    },
    {
        "id": "34433",
        "name": "34433",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34433",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34433"
            }
        ]
    },
    {
        "id": "34429",
        "name": "34429",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34429",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34429"
            }
        ]
    },
    {
        "id": "34428",
        "name": "34428",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34428",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34428"
            }
        ]
    },
    {
        "id": "34423",
        "name": "34423",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34423",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34423"
            }
        ]
    },
    {
        "id": "34146",
        "name": "34146",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34146",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34146"
            }
        ]
    },
    {
        "id": "34145",
        "name": "34145",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34145",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34145"
            }
        ]
    },
    {
        "id": "34143",
        "name": "34143",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34143",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34143"
            }
        ]
    },
    {
        "id": "34140",
        "name": "34140",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34140",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34140"
            }
        ]
    },
    {
        "id": "34139",
        "name": "34139",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34139",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34139"
            }
        ]
    },
    {
        "id": "34138",
        "name": "34138",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34138",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34138"
            }
        ]
    },
    {
        "id": "34137",
        "name": "34137",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34137",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34137"
            }
        ]
    },
    {
        "id": "34120",
        "name": "34120",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34120",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34120"
            }
        ]
    },
    {
        "id": "34117",
        "name": "34117",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34117",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34117"
            }
        ]
    },
    {
        "id": "34116",
        "name": "34116",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34116",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34116"
            }
        ]
    },
    {
        "id": "34114",
        "name": "34114",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34114",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34114"
            }
        ]
    },
    {
        "id": "34113",
        "name": "34113",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34113",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34113"
            }
        ]
    },
    {
        "id": "34112",
        "name": "34112",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34112",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34112"
            }
        ]
    },
    {
        "id": "34109",
        "name": "34109",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34109",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34109"
            }
        ]
    },
    {
        "id": "34108",
        "name": "34108",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34108",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34108"
            }
        ]
    },
    {
        "id": "34107",
        "name": "34107",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34107",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34107"
            }
        ]
    },
    {
        "id": "34106",
        "name": "34106",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34106",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34106"
            }
        ]
    },
    {
        "id": "34105",
        "name": "34105",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34105",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34105"
            }
        ]
    },
    {
        "id": "34104",
        "name": "34104",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34104",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34104"
            }
        ]
    },
    {
        "id": "34103",
        "name": "34103",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34103",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34103"
            }
        ]
    },
    {
        "id": "34102",
        "name": "34102",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34102",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34102"
            }
        ]
    },
    {
        "id": "34101",
        "name": "34101",
        "priority": 1,
        "origin": null,
        "categories": [],
        "shipping": true,
        "product_classes": [
            "standard"
        ],
        "account_classes": [],
        "rates": [
            {
                "name": "34101",
                "country": "US",
                "rate": 0,
                "state": "FL",
                "zip": "34101"
            }
        ]
    }
]

**Reply by ericingram** [reply]
Since you have the full array of rules, assuming you want to override everything existing, try this:

```
swell.put('/settings/taxes', {
  $set: {
    rules: <your_rules_array_here>
  }
})
